rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1796,10 +2438,85 @@ rmw_send_response(
-  static_cast<void>(request_header);
-  static_cast<void>(ros_response);
-  return RMW_RET_UNSUPPORTED;
+  RCUTILS_LOG_INFO_NAMED(
+    "rmw_zenoh_cpp", "[rmw_send_response]");
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(service->data, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(request_header, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    service,
+    service->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    service->data,
+    "Unable to retrieve service_data from service",
+    RMW_RET_INVALID_ARGUMENT);
+
+  rmw_service_data_t * service_data = static_cast<rmw_service_data_t *>(service->data);
+
+  rcutils_allocator_t * allocator = &(service_data->context->options.allocator);
+
+  size_t max_data_length = (
+    service_data->response_type_support->get_estimated_serialized_size(
+      ros_response, service_data->response_type_support_impl));
+
+  // Init serialized message byte array
+  char * response_bytes = static_cast<char *>(allocator->allocate(
+      max_data_length,
+      allocator->state));
+  if (!response_bytes) {
+    RMW_SET_ERROR_MSG("failed allocate response message bytes");
+    allocator->deallocate(response_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  // Object that manages the raw buffer
+  eprosima::fastcdr::FastBuffer fastbuffer(response_bytes, max_data_length);
+
+  // Object that serializes the data
+  eprosima::fastcdr::Cdr ser(
+    fastbuffer,
+    eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
+    eprosima::fastcdr::Cdr::DDS_CDR);
+  if (!service_data->response_type_support->serialize_ros_message(
+      ros_response,
+      ser,
+      service_data->response_type_support_impl))
+  {
+    allocator->deallocate(response_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  size_t data_length = ser.getSerializedDataLength();
+
+  size_t meta_length = sizeof(request_header->sequence_number);
+  memcpy(
+    &response_bytes[data_length],
+    reinterpret_cast<char *>(&request_header->sequence_number),
+    meta_length);
+
+  // Create the queryable payload
+  std::lock_guard<std::mutex> lock(service_data->query_queue_mutex);
+  auto query_it = service_data->id_query_map.find(request_header->sequence_number);
+  if (query_it == service_data->id_query_map.end()) {
+    RMW_SET_ERROR_MSG("Unable to find taken request. Report this bug.");
+    return RMW_RET_ERROR;
+  }
+  const z_query_t z_loaned_query = z_query_loan(&query_it->second);
+  z_query_reply_options_t options = z_query_reply_options_default();
+  options.encoding = z_encoding(Z_ENCODING_PREFIX_EMPTY, NULL);
+  z_query_reply(
+    &z_loaned_query, z_loan(service_data->keyexpr), reinterpret_cast<const uint8_t *>(
+      response_bytes), data_length + meta_length, &options);
```

Comment:
Do we need to free `response_bytes` after we're done?  Or is that handled by the destructor for `FastBuffer`?

Reply:
It turns out that we do need to always free response_bytes.  I've added a scope_exit for this in [7773aee](https://github.com/ros2/rmw_zenoh/pull/86/commits/7773aee1b9dbbad38c897527c25af04041c181b0)